### PR TITLE
Task_19984_86_87_89: Create UI, Handle Event, Fetch and Display data

### PR DESCRIPTION
**Pull request by khanhnn-1791**
### Issue Link :link:

https://edu-redmine.sun-asterisk.vn/issues/11984
https://edu-redmine.sun-asterisk.vn/issues/11986
https://edu-redmine.sun-asterisk.vn/issues/11987
https://edu-redmine.sun-asterisk.vn/issues/11989
### Goals :soccer:

https://marvelapp.com/be90ba4/screen/56472358
- Upcoming, Popular tab
- Search, Extend navbar button
- Movie list collectionview
- Recent watched movie list tableview

- Click to Upcoming | Popular tab to list upcoming | popular movies
- Click to a movie banner to go to Movie Detail Screen
- Click to a movie on recent watched tableview to go to Movie Detail Screen
- Click to search navbar button to go to Search Movie Screen
- Click to extend navbar button to show | hide Home Extension Screen

Fetch data and display to UI:
- Upcoming and Popular movie list

### Implementation Details :construction:
https://gfycat.com/UnhappyShimmeringHoneybadger

### CODE REVIEW

**Comment by khanhnn-1791** [comment]
ready

File: MovieApp/Scenes/GettingStarted/GettingStartedNavigator.swift [hunk 1]
```diff
@@ -20,6 +20,12 @@ final class DefaultGettingStartedNavigator: GettingStartedNavigator {
         self.viewController = viewController
     }
     
-    func toMain() {
+    func toMain() {        
+        if let vc = Storyboards.main.instantiateViewController(withIdentifier: "MainViewController")
```
Comment (tienpx3010) [inline MovieApp/Scenes/GettingStarted/GettingStartedNavigator.swift]: Sử dụng reusable

File: MovieApp/Scenes/Main/MainViewController.swift [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        output.movies
+            .drive(collectionView.rx.items) { collectionView, index, movie in
+                let indexPath = IndexPath(row: index, section: 0)
+                let cell: MovieDisplayCell = collectionView.dequeueReusableCell(for: indexPath)
+                cell.setContent(movie: movie)
+                return cell
+            }
+            .disposed(by: disposeBag)
+        
+        output.popularMovies
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewController.swift]: merge popularMovies và upcommingMovies

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            }
+            .do(onNext: { (movies) in
+                self.totalMovies.accept(movies)
+            })
+            .mapToVoid()
+            .asDriver(onErrorJustReturn: ())
+        
+        let selected = input.selectTrigger
+            .withLatestFrom(totalMovies.asDriver()) { indexPath, movies -> Driver<Void> in
+                let selectedItem = movies[indexPath.item]
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]: đưa vào do(onNext)

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        
+        let loadMore = input.loadMoreTrigger
+            .withLatestFrom(isPopular.asDriver(onErrorJustReturn: true)) { _, isPopular -> Driver<Void> in
+                if isPopular {
+                    self.popularPage += 1
+                } else {
+                    self.upcommingPage += 1
+                }
+                let result = isPopular ?
+                    self.movieRepository.getPopularMovie(input: PopularMovieRequest(page: self.popularPage))
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]: flatmapLastest

File: MovieApp/Utils/Constants.swift [hunk 5]
```diff
@@ -11,3 +11,8 @@ import UIKit
 enum Constants {
     static let appName = "MovieApp"
 }
+
+enum Cells {
```
Comment (tienpx3010) [inline MovieApp/Utils/Constants.swift]: Đưa vào MainVC

File: MovieApp/Scenes/Main/MovieListCell/MovieListCell.swift [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import UIKit
+import Reusable
+import Kingfisher
+
+final class MovieListCell: UITableViewCell, NibReusable {
+    @IBOutlet private weak var posterImageView: UIImageView!
+    @IBOutlet private weak var titleLabel: UILabel!
+    @IBOutlet private weak var ratingLabel: UILabel!
+    
+    override func awakeFromNib() {
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MovieListCell/MovieListCell.swift]: Bỏ

File: MovieApp/Scenes/Main/MovieDisplayCell/MovieDisplayCell.swift [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+            $0.layer.cornerRadius = 10
+        }
+    }
+    
+    func setContent(movie: Movie) {
+        guard let url = URL(string: "\(URLs.posterApi)\(movie.posterPath)") else { return }
+        titleLabel.text = movie.title
+        ratingLabel.text = "\(movie.voteAverage)/10"
+        let resource = ImageResource(downloadURL: url, cacheKey: movie.posterPath)
+        posterImageView.kf.setImage(with: resource)
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MovieDisplayCell/MovieDisplayCell.swift]:
```suggestion
        posterImageView.kf.setImage(with: URL(string:.....))
```

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+                } catch {
+                    return Driver.just(())
+                }
+        }
+        
+        
+        let loadMore = input.loadMoreTrigger
+            .flatMapLatest { _ -> Driver<Void> in
+                var isPopular = true
+                self.isPopular.take(1).subscribe{
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]: truyền loại popular hay upcoming từ loadMoreTrigger nhé

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+        
+        return Output(error: errorTracker.asDriver(),
+                      categoryMovies: toggle.asDriver(onErrorJustReturn: Bool()),
+                      movies: totalMovies.asDriver(),
+                      selectedCell: selected,
+                      recentlyWatchedMovies: recentlyWatchedMovies,
+                      indicator: activityIndicator.asDriver(),
+                      searchAction: searchAction,
+                      homeAction: homeExtensionAction,
+                      loaded: loaded, loadMore: loadMore)
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]:
```suggestion
                      loaded: loaded, 
                      loadMore: loadMore)
```

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+                        realmConfig.fileURL = fileURL.deletingLastPathComponent().appendingPathComponent(RealmConstansts.recentlyWatchedMovies)
+                    }
+                    let realm = try Realm(configuration: realmConfig)
+                    try realm.write {
+                        realm.add(selectedItem)
+                    }
+                    self.navigator.toDetailMovie(movie: selectedItem)
+                    return Driver.just(())
+                } catch {
+                    return Driver.just(())
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]:
```suggestion
                    return Driver.empty()
```

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+                do {
+                    var realmConfig = Realm.Configuration.defaultConfiguration
+                    if let fileURL = realmConfig.fileURL {
+                        realmConfig.fileURL = fileURL.deletingLastPathComponent().appendingPathComponent(RealmConstansts.recentlyWatchedMovies)
+                    }
+                    let realm = try Realm(configuration: realmConfig)
+                    try realm.write {
+                        realm.add(selectedItem)
+                    }
+                    self.navigator.toDetailMovie(movie: selectedItem)
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]: Bỏ vào .do(onNext

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+        self.navigator = navigator
+    }
+    
+    func transform(input: MainViewModel.Input) -> MainViewModel.Output {
+        let errorTracker = ErrorTracker()
+        let activityIndicator = ActivityIndicator()
+        
+        let popularButton = input.popularButtonTrigger
+            .do(onNext: { _ in
+                self.popularPage = 1
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]: Tại sao chuyển trang lại set lại page?

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+            .mapToVoid()
+        
+        let upcommingButton = input.upcommingButtonTrigger
+            .do(onNext: { _ in
+                self.upcommingPage = 1
+                self.isPopular.onNext(false)
+            })
+            .mapToVoid()
+        
+        let toggle = Observable.of(popularButton.map({ _ -> Bool in
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]:
```suggestion
        let toggle = Driver.merge(input.upcommingButtonTrigger.map { true }, input.popularButtonTrigger.map { false })
```

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+//  Created by nguyen.nam.khanh on 5/16/19.
+//  Copyright © 2019 nguyen.nam.khanh. All rights reserved.
+//
+
+import RealmSwift
+
+final class MainViewModel: ViewModelType {
+    private let navigator: MainNavigator
+    private let movieRepository: MovieRepository
+    private let isPopular = BehaviorSubject<Bool>(value: true)
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]: Nên set thành 2 collectionView. Khi click thì ẩn cái kia đi.

File: MovieApp/Scenes/Main/MainViewController.swift [hunk 15]
```diff
@@ -0,0 +1,131 @@
+//
+//  MainViewController.swift
+//  MovieApp
+//
+//  Created by nguyen.nam.khanh on 5/16/19.
+//  Copyright © 2019 nguyen.nam.khanh. All rights reserved.
+//
+
+enum Cells {
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewController.swift]: Bỏ trong class MainViewController

File: MovieApp/Scenes/Main/MainViewController.swift [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+            $0.layer.masksToBounds = true
+            $0.layer.cornerRadius = $0.layer.frame.width/2
+        }
+    }
+    
+    func bindViewModel() {
+        let navigator = DefaultMainNavigator(navigationController: self.navigationController)
+        viewModel =  MainViewModel(movieRepository: movieRepository, navigator: navigator)
+        
+        let input =  MainViewModel.Input(popularLoadMoreTrigger: popularLoadMoreTrigger.asDriver(onErrorJustReturn: ()), upcommingLoadMoreTrigger: upcommingLoadMoreTrigger.asDriver(onErrorJustReturn: ()),
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewController.swift]: Xuống dòng. asDriverOnErrorJustComplete

File: MovieApp/Scenes/Main/MainViewController.swift (hunk 15 above)
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewController.swift]: LayoutOptions đúng hơn

File: MovieApp/Scenes/Main/MainViewController.swift [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+            .drive()
+            .disposed(by: disposeBag)
+        
+        output.upcommingSelected
+            .drive()
+            .disposed(by: disposeBag)
+    }
+    
+    func scrollViewDidScroll(_ scrollView: UIScrollView) {
+            let offset: CGFloat = 300
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewController.swift]: Bỏ vào LayoutOptions

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+            })
+            .mapToVoid()
+
+        let popularSelected = input.popularSelectTrigger
+            .withLatestFrom(popularTotalMovies.asDriver()) { indexPath, movies in
+                return movies[indexPath.row]
+            }
+            .do(onNext: { movie in
+                self.navigator.toDetailMovie(movie: movie)
+            }).mapToVoid()
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]: .mapToVoid() xuống dòng

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+        
+        let popularLoadMore = input.popularLoadMoreTrigger
+            .flatMapLatest { _ -> Driver<[Movie]> in
+                self.popularPage += 1
+                let result = self.movieRepository.getPopularMovie(input: PopularMovieRequest(page: self.popularPage))
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]:
```suggestion
                return self.movieRepository.getPopularMovie(input: PopularMovieRequest(page: self.popularPage))
```

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+//
+
+import RealmSwift
+
+final class MainViewModel: ViewModelType {
+    private let navigator: MainNavigator
+    private let movieRepository: MovieRepository
+    private let popularTotalMovies = BehaviorRelay<[Movie]>(value: [])
+    private let upcommingTotalMovies = BehaviorRelay<[Movie]>(value: [])
+    private var upcommingPage = 1
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]: BehaviorRelay

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+            }.do(onNext: { movies in
+                var current = self.upcommingTotalMovies.value
+                current += movies
+                self.upcommingTotalMovies.accept(current)
+            })
+            .mapToVoid()
+            .asDriver(onErrorJustReturn: ())
+        
+        let searchAction = input.searchButtonTrigger
+            .do(onNext: { _ in
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]:
```suggestion
            .do(onNext: navigator.toSearch
```

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+            }
+            .do(onNext: { movie in
+                self.navigator.toDetailMovie(movie: movie)
+            })
+            .mapToVoid()
+        
+        let popularLoadMore = input.popularLoadMoreTrigger
+            .throttle(1)
+            .flatMapLatest { _ -> Driver<[Movie]> in
+                self.popularPage.accept(self.popularPage.value + 1)
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift:79]: Add do() ở trước flatMapLatest

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+                    .trackError(errorTracker)
+                    .asDriverOnErrorJustComplete()
+            }
+            .do(onNext: { (movies) in
+                self.upcommingTotalMovies.accept(movies)
+            })
+            .mapToVoid()
+
+        let popularSelected = input.popularSelectTrigger
+            .withLatestFrom(popularTotalMovies.asDriver()) { indexPath, movies in
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]: Tạo 1 biến = popularTotalMovies.asDriver()

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 24]
```diff
@@ -0,0 +1,150 @@
...
+                    .asDriverOnErrorJustComplete()
+            }
+            .do(onNext: { (movies) in
+                self.popularTotalMovies.accept(movies)
+            })
+            .mapToVoid()
+        
+        let loadUpcomming = input.loaded
+            .flatMapLatest { _ in
+                return self.movieRepository.getUpcommingMovie(input: UpcommingMovieRequest(page: self.upcommingPage.value))
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]: withLatestFrom upcommingPage

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+        self.navigator = navigator
+    }
+    
+    func transform(input: MainViewModel.Input) -> MainViewModel.Output {
+        let errorTracker = ErrorTracker()
+        let activityIndicator = ActivityIndicator()
+
+        let loadPopular = input.loaded
+            .flatMapLatest { _ in
+                return self.movieRepository.getPopularMovie(input: PopularMovieRequest(page: self.popularPage.value))
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]: withLatestFrom

File: MovieApp/Scenes/Main/MainViewController.swift [hunk 26]
```diff
@@ -0,0 +1,132 @@
+//
+//  MainViewController.swift
+//  MovieApp
+//
+//  Created by nguyen.nam.khanh on 5/16/19.
+//  Copyright © 2019 nguyen.nam.khanh. All rights reserved.
+//
+
+final class MainViewController: UIViewController {
+    enum LayoutOptions {
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewController.swift]: private

File: MovieApp/Scenes/Main/MainViewController.swift [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+    @IBOutlet private weak var collectionView: UICollectionView!
+    @IBOutlet private weak var logoImageView: UIImageView!
+    @IBOutlet private weak var searchButton: UIButton!
+    @IBOutlet private weak var homeExtensionButton: UIButton!
+    @IBOutlet private weak var upcommingCollectionView: UICollectionView!
+    
+    private let disposeBag = DisposeBag()
+    private var viewModel: MainViewModel!
+    private let movieRepository: MovieRepository  = MovieRepositoryImpl(api: APIService.share)
+    let popularLoadMoreTrigger = PublishSubject<Void>()
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewController.swift]: private

File: MovieApp/Scenes/Main/MainViewModel.swift (hunk 24 above)
Comment (tienpx3010) [reply MovieApp/Scenes/Main/MainViewModel.swift]:
```suggestion
                return self.movieRepository.getUpcommingMovie(input: UpcommingMovieRequest(page: page))
```

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 28]
```diff
@@ -0,0 +1,160 @@
...
+    }
+    
+    func transform(input: MainViewModel.Input) -> MainViewModel.Output {
+        let errorTracker = ErrorTracker()
+        let activityIndicator = ActivityIndicator()
+        let popularMoviesDriver = popularTotalMovies.asDriver()
+        let upcommingMoviesDriver = upcommingTotalMovies.asDriver()
+
+        let loadPopular = input.loaded
+            .withLatestFrom(popularPage.asDriver()) { _, popularPage -> Int in
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]:
```suggestion
            .withLatestFrom(popularMoviesDriver)
```

File: MovieApp/Scenes/Main/MainViewModel.swift [hunk 29]
```diff
@@ -0,0 +1,160 @@
...
+                    .trackError(errorTracker)
+                    .asDriverOnErrorJustComplete()
+            }
+            .do(onNext: { (movies) in
+                self.popularTotalMovies.accept(movies)
+            })
+            .mapToVoid()
+        
+        let loadUpcomming = input.loaded
+            .withLatestFrom(upcommingPage.asDriver()) { _, upcommingPage -> Int in
```
Comment (tienpx3010) [inline MovieApp/Scenes/Main/MainViewModel.swift]:
```suggestion
            .withLatestFrom(upcommingMoviesDriver)
```

**Review by tienpx3010 — APPROVED** [review]

**Pull request merged**